### DISCUSSION TITLE
User modifications with Yosys

### DISCUSSION THREAD

**Discussion by gudeh** [The-OpenROAD-Project/OpenROAD-flow-scripts#357] [Q&A] [answered]
Hello everyone! I am trying to use the [OpenROAD-flow and Yosys to generate some CSV data](https://github.com/The-OpenROAD-Project/OpenROAD/discussions/1610). The idea is to run Yosys with small modifications to output the data, afterwards run the OpenROAD-flow and also generate some CSV data. The OpenROAD-flow part is working fine. Although I am having some issues with Yosys.

At first I intended to run Yosys outside of OpenROAD, on a separte project and afterwards run the flow (seems easier). This is not an option because OpenROAD seems to generate its own .lib library, so I am obligated to run Yosys via the OpenROAD-flow.  Do you agree with this, or is it an option to get a generated .lib from the flow so I can use it on Yosys myself? Either way this doesn't seem like the proper solution.

- A better solution is to edit the Yosys project being built by the OpenROAD-flow. I tried to do so [in my fork](https://github.com/gudeh/OpenROAD-flow-scripts), although I am having some issues. I don't know how to properly edit a submodule in a way that ./build_openroad.sh would still work. The only way I could make the submodule download my modifications (even with the [.gitmodules file pointing to my Yosys fork](https://github.com/gudeh/OpenROAD-flow-scripts/blob/master/.gitmodules)) was with the command `sudo git submodule foreach git pull origin master`. Anyway, the real issue is when I run the build script with `./build_openroad.sh --local --latest`, it seems to point to OpenROAD's git URL for Yosys, since it override my modifications.

I hope I was clear enough, buy in summary I want to build the OpenROAD-flow with my modifications on Yosys!

**Comment by maliberty** [comment, marked as answer]
I suggest you fork yosys and ORFS.  In your ORFS you can update the submodule to point to your version of yosys with

git submodule set-url [--] <path> <newurl>

**Reply by gudeh** [reply]
Didn't work with this command either

**Reply by gudeh** [reply]
I already have my own forks for ORFS and Yosys, and Yosys already points to my fork in `.gitmodules` file.

**Reply by maliberty** [reply]
Where is your fork?

**Reply by gudeh** [reply]
ORFS: https://github.com/gudeh/OpenROAD-flow-scripts
Yosys: https://github.com/gudeh/yosys

**Reply by vvbandeira** [reply]
@gudeh 
To update the URLs `git submodule sync`.
To skip submodule update during ORFS build, add the argument `-no_init` to the `build_openroad.sh`.

**Reply by maliberty** [reply]
I suspect you have an existing area where the submodules are already checked out from the old path.  If you clone your ORFS in a fresh area do you still see the issue?

**Reply by gudeh** [reply]
I was only missing the `--no_init`.

**Reply by gudeh** [reply]
> I suspect you have an existing area where the submodules are already checked out from the old path. If you clone your ORFS in a fresh area do you still see the issue?

Actually, I tried a couple of times from a clean `git clone` and the only way I could make it work was with this two commands:
```
sudo git submodule foreach git pull origin master
./build_openroad.sh --local --latest --no_init
```
without the first one it just compiles Yosys without my fork. Without saying `.gitmodules` is always linked to my fork.

**Reply by gudeh** [reply]
Hey guys, this is not working anymore and I can't figure out why.
When I try to compile everything with 'build_openroad.sh', including my Yosys fork, I see a bunch of this warnings during Yosys build:
`cc1: warning: command-line option ‘-fpermissive’ is valid for C++/ObjC++ but not for C`

Any idea what might be happening? My Yosys fork points to ORFS's Yosys fork and is up to date: https://github.com/gudeh/yosys

**Reply by maliberty** [reply]
what compiler/version are you using?

**Reply by gudeh** [reply]
I noticed this difference when building with my Yosys fork:
`[ 94%] ABC: Using CFLAGS=-Wall -Wno-unused-function -Wno-write-strings -Wno-sign-compare -DABC_USE_STDINT_H  -DABC_NAMESPACE=abc -fpermissive -DABC_USE_CUDD=1 -DABC_USE_READLINE  -DABC_USE_PTHREADS -Wno-unused-but-set-variable`

From the standar ORFS:
`Using CFLAGS=-Wall -Wno-unused-function -Wno-write-strings -Wno-sign-compare -DLIN64 -DSIZEOF_VOID_P=8 -DSIZEOF_LONG=8 -DSIZEOF_INT=4 -DABC_USE_CUDD=1 -DABC_USE_READLINE  -DABC_USE_PTHREADS -Wno-unused-but-set-variable`

**Reply by gudeh** [reply]
I am trying to compile from my fork of ORFS, which is also up to date with ORFS: https://github.com/gudeh/OpenROAD-flow-scripts

**Reply by gudeh** [reply]
And this is the build log:
[build_openroad.log](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/files/12785188/build_openroad.log)

**Reply by gudeh** [reply]
The error seems to be when compiling ABC. I noticed it is using the correct version of ABC (the same as ORFS standard): `Building abc/abc-bafd2a7`

Although, the CFLAGS are different.

This is awkward to me since I have only a new pass on Yosys fork, and two new tcl scrips in ORFS fork, I don't understand how it generates a different CFLAGS.

**Reply by vvbandeira** [reply]
@gudeh 
If you don't have any changes to `abc`, try removing the folder and compiling again:
```
rm -rf tools/yosys/abc
```
Or
```
./build_openroad.sh --clean-force [ARGS]
```

**Reply by gudeh** [reply]
Same error, unfortunately.

**Comment by rovinski** [comment]
> OpenROAD seems to generate its own .lib library

The reason for this is because Yosys does not have any equivalent of a `set_dont_use` command. The only way to tell Yosys not to use cells is to edit the `dont_use` property directly in the `.lib` file which is read by Yosys. To do this, OpenROAD-flow creates a temporary copy and then edits the properties there to keep the original `.lib` intact.

**Reply by gudeh** [reply]
Hi @rovinski, what if I wanted to use the same library as the one used during the OR flow? Is it safe to read the one in "::env(DONT_USE_LIBS)" ?

**Reply by rovinski** [reply]
Yes. Those libraries simply have some of the `dont_use` properties set in accordance with what's set in the platform / user config file. You can `diff` the `DONT_USE_LIBS` with the ones in `PLATFORM_DIR` to see for yourself.

As a side note, ABC and Yosys recently added support for setting `dont_use` properties via script. Once (if) that version is upstreamed to ORFS, I believe we will simply remove the `DONT_USE_LIBS` because they will be obsolete.

**Reply by gudeh** [reply]
Great! Thanks a lot for clarifying.
